airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/FlushWorkers.java

```diff
@@ -53,93 +54,132 @@ public class FlushWorkers implements AutoCloseable {
     this.bufferDequeue = bufferDequeue;
-    this.flusher = flushFunction;
+    flusher = flushFunction;
     this.outputRecordCollector = outputRecordCollector;
     this.flushFailure = flushFailure;
+    streamToInProgressWorkers = new ConcurrentHashMap<>();
+    debugLoop = Executors.newSingleThreadScheduledExecutor();
+    supervisorThread = Executors.newScheduledThreadPool(1);
+    workerPool = Executors.newFixedThreadPool(5);
+    isClosing = new AtomicBoolean(false);
   }
 
   public void start() {
-    supervisorThread.scheduleAtFixedRate(this::retrieveWork, SUPERVISOR_INITIAL_DELAY_SECS, SUPERVISOR_PERIOD_SECS,
+    supervisorThread.scheduleAtFixedRate(this::retrieveWork,
+        SUPERVISOR_INITIAL_DELAY_SECS,
+        SUPERVISOR_PERIOD_SECS,
+        TimeUnit.SECONDS);
+    debugLoop.scheduleAtFixedRate(this::printWorkerInfo,
+        DEBUG_INITIAL_DELAY_SECS,
+        DEBUG_PERIOD_SECS,
         TimeUnit.SECONDS);
-    debugLoop.scheduleAtFixedRate(this::printWorkerInfo, DEBUG_INITIAL_DELAY_SECS, DEBUG_PERIOD_SECS, TimeUnit.SECONDS);
   }
 
   @Override
   public void close() throws Exception {
-    flushAll();
+    log.info("Closing flush workers -- waiting for all buffers to flush");
+    isClosing.set(true);
+    // wait for all buffers to be flushed.
+    while (true) {
+      final boolean anyRecordsLeft = bufferDequeue.getBufferedStreams()
+          .stream()
+          .anyMatch(streamDesc -> bufferDequeue.getQueueSizeInRecords(streamDesc).map(size -> size > 0).orElseThrow());
+      log.info("anyRecordsLeft: {}", anyRecordsLeft);
+      if (!anyRecordsLeft) {
+        break;
+      }
+
+      log.info("Waiting for all streams to flush.");
+      Thread.sleep(1000);
+    }
+    log.info("Closing flush workers -- all buffers flushed");
 
     supervisorThread.shutdown();
     final var supervisorShut = supervisorThread.awaitTermination(5L, TimeUnit.MINUTES);
-    log.info("Supervisor shut status: {}", supervisorShut);
+    log.info("Closing flush workers -- Supervisor shut status: {}", supervisorShut);
 
-    log.info("Starting worker pool shutdown..");
+    log.info("Closing flush workers -- Starting worker pool shutdown..");
     workerPool.shutdown();
     final var workersShut = workerPool.awaitTermination(5L, TimeUnit.MINUTES);
-    log.info("Workers shut status: {}", workersShut);
+    log.info("Closing flush workers -- Workers shut status: {}", workersShut);
 
     debugLoop.shutdownNow();
   }
 
   private void retrieveWork() {
-    // todo (cgardens) - i'm not convinced this makes sense. as we get close to the limit, we should
-    // flush more eagerly, but "flush all" is never a particularly useful thing in this world.
-    // if the total size is > n, flush all buffers
-    if (bufferDequeue.getTotalGlobalQueueSizeBytes() > BufferManager.TOTAL_QUEUES_MAX_SIZE_LIMIT_BYTES) {
-      flushAll();
-      return;
+    try {
+      log.info("retrieving queues to process");
```

Comment:
```suggestion
      log.debug("retrieving queues to process");
```

Reply:
might get noisy

Reply:
yeah. we should make a pass at the logs to fix levels, but for right now i'm going to leave it. as we debug what's broken it's helpful.